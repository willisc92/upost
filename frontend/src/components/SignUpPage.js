import React from "react";
import { SingleDatePicker } from "react-dates";
import moment from "moment";
import API from "../utils/API";
import emailValidator from "email-validator";

class SignUpPage extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            calendarFocused: false,
            date: moment(),
            birthDate: moment(),
            isLoaded: false,
            interests: [],
            error: undefined
        };
    }

    componentDidMount() {
        API.get("interests/").then(
            (result) => {
                this.setState({
                    isLoaded: true,
                    interests: result.data
                });
            },

            (error) => {
                this.setState({
                    isLoaded: true,
                    error
                });
            }
        );
    }

    onDateChange = (birthDate) => {
        if (birthDate) {
            this.setState(() => ({ birthDate }));
        }
    };

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    };

    onSubmit = (e) => {
        e.preventDefault();
        const input = e.target;

        // validate email
        if (!emailValidator.validate(input.email.value)) {
            alert("invalid email");
        }

        // check password
        if (input.password.value !== input.passwordConfirmation.value) {
            alert("passwords don't match");
        }

        const user = {
            password: input.password.value,
            username: input.username.value,
            email: input.email.value,
            first_name: input.firstName.value,
            middle_name: input.middleName.value,
            last_name: input.lastName.value,
            birth_date: this.state.date.format("YYYY-MM-DD"),
            country: input.country.value,
            state: input.state.value,
            street_name: input.streetName.value,
            postal_code: input.postalCode.value,
            city: input.city.value,
            sex: input.sex.value === "Male" ? "M" : "F",
            phone_number: input.phoneNumber.value,
            interests: [input.interests.value]
        };

        API.post("accounts/", user)
            .then((response) => {
                console.log(response);
            })
            .catch((error) => {
                console.log("error in account creation", error);
                console.log(error.response.request.response);
                this.setState({ error: error.response.request.response });
            });

        console.log(user);
    };

    render() {
        return (
            <form onSubmit={this.onSubmit}>
                <p>Username*</p>
                {!!this.state.error.username && <p>{this.state.error.username}</p>}
                <input type="text" name="username" />
                <p>150 characters of fewer. Letters, digits and @/./+/-/) only.</p>
                <p>Email Address</p>
                <input type="text" name="email" />
                <p>First Name*</p>
                <input type="text" name="firstName" />
                <p>Middle Name*</p>
                <input type="text" name="middleName" />
                <p>Last Name*</p>
                <input type="text" name="lastName" />
                <p>Birth date</p>
                <SingleDatePicker
                    date={this.state.date}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={(day) => false}
                />
                <p>Country*</p>
                <input type="text" name="country" />
                <p>State*</p>
                <input type="text" name="state" />
                <p>Steet Name*</p>
                <input type="text" name="streetName" />
                <p>Postal Code*</p>
                <input type="text" name="postalCode" />
                <p>City*</p>
                <input type="text" name="city" />
                <p>Sex</p>
                <select name="sex">
                    <option>Male</option>
                    <option>Female</option>
                </select>
                <p>Phone Number</p>
                <input type="text" name="phoneNumber" />
                <p>Interests*</p>
                <select name="interests">
                    {this.state.interests.map((interest) => {
                        return <option key={interest.interest_tag}>{interest.interest_tag}</option>;
                    })}
                </select>
                <p>Password</p>
                <input type="text" name="password" />
                <p>Password Confirmation*</p>
                <input type="text" name="passwordConfirmation" />
                <br />
                <button>Sign Up</button>
            </form>
        );
    }
}

export default SignUpPage;
