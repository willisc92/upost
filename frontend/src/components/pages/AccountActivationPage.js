import React from "react";
import { startActivation } from "../../actions/activate";

class AccountActivationPage extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            inProgress: true,
            isValid: undefined
        };
    }

    componentDidMount() {
        const uid = this.props.match.params.uid;
        const token = this.props.match.params.token;
        startActivation(uid, token)
            .then((result) => {
                this.setState(() => {
                    return { inProgress: false, isValid: true };
                });
            })
            .catch((error) => {
                console.log("error", JSON.stringify(error, null, 2));
                this.setState(() => {
                    return { inProgress: false, isValid: false };
                });
            });
    }

    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        {this.state.inProgress ? (
                            <h1 className="page-header__title">
                                Your <span>UPost</span> account is currently being activated!
                            </h1>
                        ) : this.state.isValid ? (
                            <h1 className="page-header__title">
                                Thank you for your email confirmation. Please login and complete your registration.
                            </h1>
                        ) : (
                            <h1 className="page-header__title">The activation link is invalid</h1>
                        )}
                    </div>
                </div>
            </div>
        );
    }
}

export default AccountActivationPage;
