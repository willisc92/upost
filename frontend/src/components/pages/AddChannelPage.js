import React from "react";
import ChannelForm from "../forms/ChannelForm";
import { connect } from "react-redux";
import { addChannel } from "../../actions/channels";

class AddChannelPage extends React.Component {
    constructor(props) {
        super(props);
    }

    onSubmit = (channel) => {
        this.props
            .addChannel({ ...channel })
            .then(() => this.props.history.push("/myChannels"))
            .catch(() => {});
    };

    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Add a Channel</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ChannelForm onSubmit={this.onSubmit} />
                </div>
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    addChannel: (channel) => dispatch(addChannel(channel))
});

export default connect(
    null,
    mapDispatchToProps
)(AddChannelPage);
