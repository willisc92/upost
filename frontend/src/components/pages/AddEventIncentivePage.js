import React from "react";
import { addIncentivePackage } from "../../actions/incentivePackage";
import { getCurrentUser } from "../../actions/auth";
import IncentiveForm from "../forms/IncentiveForm";
import { connect } from "react-redux";
import { startSetEvent, clearEvents } from "../../actions/events";

import Box from "@material-ui/core/Box";
import Container from "@material-ui/core/Container";
import Typography from "@material-ui/core/Typography";
import Button from "@material-ui/core/Button";

class AddEventIncentivePage extends React.Component {
    constructor(props) {
        super(props);
    }

    componentWillMount() {
        this.props.clearEvents();
        const event_id = this.props.match.params.id;
        getCurrentUser()
            .then((res) => {
                this.props
                    .startSetEvent(event_id)
                    .then((event_res) => {
                        if (res.data.username !== event_res.data.event_owner) {
                            this.props.history.push(`/myChannels`);
                        }
                    })
                    .catch((err) => {
                        console.log(JSON.stringify(err, null, 2));
                    });
            })
            .catch((err) => {
                console.log(JSON.stringify(err, null, 2));
            });
    }

    onSubmit = (incentive) => {
        const post_id = this.props.event.post;
        this.props
            .addIncentivePackage(incentive)
            .then((res) => this.props.history.push(`/post-events/${post_id}`))
            .catch((err) => {
                console.log(err);
            });
    };

    goBack = () => {
        const event_id = this.props.match.params.id;
        const post_id = this.props.event.post;
        this.props.history.push(`/myPosts/${post_id}/events/${event_id}/edit`);
    };

    goToIncentive = () => {
        const event_id = this.props.match.params.id;
        this.props.history.push(`/myEvents/${event_id}/editIncentive`);
    };

    render() {
        const read_only = !!this.props.event && this.props.event.deleted_flag;
        const existing_incentive = !!this.props.event && !!this.props.event.event_incentive;

        return (
            !!this.props.event && (
                <div>
                    <Box bgcolor="secondary.main" py={3}>
                        <Container fixed>
                            <Typography gutterBottom variant="h2">
                                Add an Incentive Package to Event:{" "}
                                <Typography variant="inherit" display="inline" color="primary">
                                    {this.props.event && this.props.event.event_title}
                                </Typography>
                            </Typography>

                            {existing_incentive ? (
                                <Box paddingBottom={2}>
                                    <Typography variant="h2" color="error" gutterBottom>
                                        There is already an existing incentive tied to this event.
                                    </Typography>
                                    <Button variant="contained" color="primary" onClick={this.goToIncentive}>
                                        Go to Incentive
                                    </Button>
                                </Box>
                            ) : (
                                read_only && (
                                    <Typography variant="h2" color="error" gutterBottom>
                                        The event this incentive will be tied to is deleted. Restore it before adding an
                                        incentive.
                                    </Typography>
                                )
                            )}
                            <Box>
                                <Button variant="contained" color="primary" onClick={this.goBack}>
                                    Go to Event
                                </Button>
                            </Box>
                        </Container>
                    </Box>
                    <Container fixed>
                        <IncentiveForm
                            onSubmit={this.onSubmit}
                            nextStep="Save"
                            event={this.props.event}
                            fromEvent={true}
                            read_only={read_only || existing_incentive}
                        />
                    </Container>
                </div>
            )
        );
    }
}

const mapStateToProps = (state) => ({
    event: state.events.events.length !== 0 ? state.events.events : null
});

const mapDispatchToProps = (dispatch) => ({
    clearEvents: () => dispatch(clearEvents()),
    addIncentivePackage: (incentive) => dispatch(addIncentivePackage(incentive)),
    startSetEvent: (id) => dispatch(startSetEvent(id))
});

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(AddEventIncentivePage);
