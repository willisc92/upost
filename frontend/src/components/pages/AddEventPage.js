import React from "react";
import { addEvent } from "../../actions/events";
import { getCurrentUser } from "../../actions/auth";
import EventForm from "../forms/EventForm";
import IncentiveForm from "../forms/IncentiveForm";
import { connect } from "react-redux";
import { startGetPost, clearPosts } from "../../actions/posts";
import { addIncentivePackage } from "../../actions/incentivePackage";
import { RecurringModal } from "../modals/RecurringModal";
import { mapFrequencyToRRule, mapDayToRRule } from "../../utils/recurring";
import { RRule } from "rrule";
import MessageModal from "../modals/MessageModal";
import CustomStepper from "../CustomStepper";

import Box from "@material-ui/core/Box";
import Container from "@material-ui/core/Container";
import Typography from "@material-ui/core/Typography";
import Button from "@material-ui/core/Button";

class AddEventPage extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            messageOpen: false,
            message: "",
            step: "Event",
            base_event: null,
            base_incentive: null,
            finished: false,
            recurringOpen: false,
            rrule_payload: {
                recurringFrequency: "none",
                endSelection: "none",
                numOccurences: 0,
                endDate: null,
                byweekday: null
            },
            rrule_event_starts: null,
            rrule_event_ends: null,
            rrule_incentive_starts: null,
            rrule_incentive_ends: null,
            error: "",
            steps: [],
            activeStep: undefined
        };
    }

    componentWillMount() {
        this.props.clearPosts();
        const post_id = this.props.match.params.id;
        getCurrentUser()
            .then((res) => {
                this.props
                    .startGetPost(post_id)
                    .then((post_res) => {
                        if (res.data.username !== post_res.data[0].user) {
                            this.props.history.push(`/myChannels`);
                        } else {
                            this.setState(() => ({
                                steps: [
                                    { label: "Bulletin Boards", onClick: this.moveToBulletinBoards },
                                    {
                                        label: `Bulletin Board`,
                                        onClick: this.goToChannel
                                    },
                                    { label: `Post: ${this.props.post.post_title}`, onClick: this.returnToPost },
                                    { label: "Events", onClick: this.goBack },
                                    { label: "Add Event", onClick: this.submitFormAndSwitch },
                                    { label: "Add Perks (Optional)", onClick: this.submitFormAndSwitch }
                                ],
                                activeStep: 4
                            }));
                        }
                    })
                    .catch((err) => {
                        console.log(JSON.stringify(err, null, 2));
                    });
            })
            .catch((err) => {
                console.log(JSON.stringify(err, null, 2));
            });
    }

    toggleForms = async () => {
        this.setState({
            step: "Event"
        });
    };

    goBack = () => {
        const post_id = this.props.match.params.id;
        this.props.history.push(`/post-events/${post_id}`);
    };

    goToPost = () => {
        const post_id = this.props.match.params.id;
        this.props.history.push(`/myPosts/${post_id}/edit`);
    };

    recurringSubmit = async (payload) => {
        await this.setState(() => ({
            rrule_payload: payload,
            recurringOpen: false
        }));

        this.generateEventRule(payload);
        this.generateIncentiveRule(payload);
    };

    recurringOpen = () => {
        this.setState(() => ({
            recurringOpen: true
        }));
    };

    clearBaseIncentive = () => {
        this.setState(() => ({
            base_incentive: null,
            step: "Event",
            rrule_incentive_starts: null,
            rrule_incentive_ends: null,
            message_open: true,
            message: "Perks Cleared"
        }));
    };

    clearBaseEvent = () => {
        this.setState(() => ({
            base_event: null,
            step: "Event",
            rrule_event_starts: null,
            rrule_event_ends: null,
            message_open: true,
            message: "Event Cleared"
        }));
    };

    clearBaseEventandIncentive = () => {
        this.setState(() => ({
            base_event: null,
            step: "Event",
            rrule_event_starts: null,
            rrule_event_ends: null,
            base_incentive: null,
            rrule_incentive_starts: null,
            rrule_incentive_ends: null,
            message_open: true,
            message: "Event and Perks Cleared"
        }));
    };

    generateEventRule = (payload) => {
        if (payload.recurringFrequency === "none") {
            this.setState(() => ({
                rrule_event_starts: null,
                rrule_event_ends: null
            }));
        }

        if (!!this.state.base_event && payload.endSelection === "numOccurence") {
            this.setState(() => ({
                event_error: "",
                rrule_event_starts: new RRule({
                    freq: mapFrequencyToRRule(payload.recurringFrequency),
                    dtstart: this.state.base_event.planned_start_date,
                    count: payload.numOccurences
                }),
                rrule_event_ends: new RRule({
                    freq: mapFrequencyToRRule(payload.recurringFrequency),
                    dtstart: this.state.base_event.planned_end_date,
                    count: payload.numOccurences
                })
            }));
        }

        if (!!this.state.base_event && payload.endSelection === "onDate") {
            if (this.state.base_event.planned_end_date > payload.endDate) {
                this.setState(() => ({
                    event_error:
                        "Error, cannot have a recurring end date before the event end date.  Change either the recurring end date or the end date of the event."
                }));
                return null;
            } else {
                payload.byweekday.map((day) => {
                    return mapDayToRRule(day);
                });

                this.setState(() => ({
                    event_error: "",
                    rrule_event_starts: new RRule({
                        freq: mapFrequencyToRRule(payload.recurringFrequency),
                        dtstart: this.state.base_event.planned_start_date,
                        until: payload.endDate,
                        byweekday: payload.byweekday.length === 0 ? null : payload.byweekday
                    }),
                    rrule_event_ends: new RRule({
                        freq: mapFrequencyToRRule(payload.recurringFrequency),
                        dtstart: this.state.base_event.planned_end_date,
                        until: payload.endDate,
                        byweekday: payload.byweekday.length === 0 ? null : payload.byweekday
                    })
                }));
            }
        }
    };

    generateIncentiveRule = (payload) => {
        if (!!this.state.base_incentive && payload.recurringFrequency !== "none") {
            this.setState(() => ({
                rrule_incentive_starts: null,
                rrule_incentive_ends: null
            }));
        }

        if (!!this.state.base_incentive && payload.endSelection === "numOccurence") {
            this.setState(() => ({
                incentive_error: "",
                rrule_incentive_starts: new RRule({
                    freq: mapFrequencyToRRule(payload.recurringFrequency),
                    dtstart: this.state.base_incentive.planned_start_date,
                    count: payload.numOccurences
                }),
                rrule_incentive_ends: new RRule({
                    freq: mapFrequencyToRRule(payload.recurringFrequency),
                    dtstart: this.state.base_incentive.planned_end_date,
                    count: payload.numOccurences
                })
            }));
        }

        if (!!this.state.base_incentive && payload.endSelection === "onDate") {
            if (this.state.base_incentive.planned_end_date > payload.endDate) {
                this.setState(() => ({
                    incentive_error:
                        "Error, cannot have a recurring end date before the perks end date.  Change either the recurring end date or the end date of the perks."
                }));
                return null;
            } else {
                payload.byweekday.map((day) => {
                    return mapDayToRRule(day);
                });
                this.setState(() => ({
                    incentive_error: "",
                    rrule_incentive_starts: new RRule({
                        freq: mapFrequencyToRRule(payload.recurringFrequency),
                        dtstart: this.state.base_incentive.planned_start_date,
                        until: payload.endDate,
                        byweekday: payload.byweekday.length === 0 ? null : payload.byweekday
                    }),
                    rrule_incentive_ends: new RRule({
                        freq: mapFrequencyToRRule(payload.recurringFrequency),
                        dtstart: this.state.base_incentive.planned_end_date,
                        until: payload.endDate,
                        byweekday: payload.byweekday.length === 0 ? null : payload.byweekday
                    })
                }));
            }
        }
    };

    submitAll = async () => {
        await this.setState(() => ({
            finished: true
        }));

        this.submitButtonRef.click();
    };

    submitWithoutSaving = async () => {
        await this.setState(() => ({
            step: "Event",
            finished: true
        }));

        this.submitButtonRef.click();
    };

    submitFormAndSwitch = () => {
        const e = new Event("submit", { cancelable: true });

        if (this.state.step === "Event") {
            document.getElementById("Event").dispatchEvent(e);
        } else {
            document.getElementById("Incentive").dispatchEvent(e);
        }
    };

    mapRRulesToText = () => {
        let rrule_event_string = null;
        let rrule_incentive_string = null;
        if (!!this.state.rrule_event_starts) {
            rrule_event_string = this.state.rrule_event_starts.toText();
        }
        if (!!this.state.rrule_incentive_starts) {
            rrule_incentive_string = this.state.rrule_incentive_starts.toText();
        }

        if (!!rrule_event_string && !rrule_incentive_string) {
            return "Events will reoccur ".concat(rrule_event_string);
        }

        if (!!rrule_event_string && !!rrule_incentive_string && rrule_event_string === rrule_incentive_string) {
            return "Events and perks will reoccur ".concat(rrule_event_string);
        }

        if (!!rrule_event_string && !!rrule_incentive_string && rrule_event_string !== rrule_incentive_string) {
            return "Events will reoccur "
                .concat(rrule_event_string)
                .concat(".  Perks will reoccur ")
                .concat(rrule_incentive_string);
        }

        return null;
    };

    handleRecurringEventSubmit = (data) => {
        const promises = [];

        for (let i = 0; i < this.state.rrule_event_starts.all().length; i++) {
            const new_payload = {
                ...data,
                planned_start_date: this.state.rrule_event_starts.all()[i],
                planned_end_date: this.state.rrule_event_ends.all()[i]
            };

            promises.push(this.props.addEvent(new_payload));
        }

        Promise.all(promises)
            .then(() => {
                this.goBack();
            })
            .catch((err) => {
                console.log(JSON.stringify(err, null, 2));
            });
    };

    handleRecurringEventandIncentiveSubmit = () => {
        for (let i = 0; i < this.state.rrule_event_starts.all().length; i++) {
            const new_event_payload = {
                ...this.state.base_event,
                planned_start_date: this.state.rrule_event_starts.all()[i],
                planned_end_date: this.state.rrule_event_ends.all()[i]
            };

            this.props
                .addEvent(new_event_payload)
                .then((res) => {
                    const event_id = res.data.event_id;
                    const new_incentive_payload = {
                        ...this.state.base_incentive,
                        planned_start_date: this.state.rrule_incentive_starts.all()[i],
                        planned_end_date: this.state.rrule_incentive_ends.all()[i],
                        event: event_id
                    };
                    this.props
                        .addIncentivePackage(new_incentive_payload)
                        .then((res) => {})
                        .catch((err) => {
                            this.setState(() => ({
                                messageOpen: true,
                                message: "Something went wrong.",
                                finished: false
                            }));
                            console.log(JSON.stringify(err, null, 2));
                            return;
                        });
                })
                .catch((err) => {
                    this.setState(() => ({
                        messageOpen: true,
                        message: "Something went wrong.",
                        finished: false
                    }));
                    console.log(JSON.stringify(err, null, 2));
                    return;
                });
        }

        this.goBack();
    };

    onSubmit = async (data) => {
        if (this.state.step === "Event") {
            await this.setState(() => ({
                base_event: data
            }));
            this.generateEventRule(this.state.rrule_payload);

            if (this.state.finished === false) {
                this.setState((prevState) => ({
                    messageOpen: true,
                    message: "Event Draft Saved",
                    step: "Incentive",
                    activeStep: prevState.activeStep + 1
                }));
            }
        } else if (this.state.step === "Incentive") {
            await this.setState(() => ({
                base_incentive: data
            }));
            this.generateIncentiveRule(this.state.rrule_payload);

            if (this.state.finished === false) {
                this.setState((prevState) => ({
                    messageOpen: true,
                    message: "Perk Package Draft Saved",
                    step: "Event",
                    activeStep: prevState.activeStep - 1
                }));
            }
        }

        if (!!this.state.finished) {
            if (!this.state.base_event) {
                await this.setState(() => ({
                    finished: false,
                    messageOpen: true,
                    message: "Must add an event prior to submission"
                }));
                return;
            }

            // Handle Single Event Submit
            if (!this.state.rrule_event_starts) {
                this.props
                    .addEvent(this.state.base_event)
                    .then((result) => {
                        if (!!this.state.base_incentive) {
                            let payload = {
                                ...this.state.base_incentive,
                                event: result.data.event_id
                            };
                            this.props
                                .addIncentivePackage(payload)
                                .then((result) => {
                                    this.goBack();
                                })
                                .catch((err) => {
                                    this.setState(() => ({
                                        finished: false
                                    }));
                                    console.log(JSON.stringify(err, null, 2));
                                });
                        } else {
                            this.goBack();
                        }
                    })
                    .catch((err) => {
                        this.setState(() => ({
                            finished: false
                        }));
                        console.log(JSON.stringify(err, null, 2));
                    });
                return;
            }

            // Handle Recurring Event Submit w/o Incentives
            if (!!this.state.rrule_event_starts && !this.state.rrule_incentive_starts) {
                this.handleRecurringEventSubmit(this.state.base_event);
            }

            // Handle Recurring Event Submit w/ Incentives
            if (!!this.state.rrule_event_starts && !!this.state.rrule_incentive_starts) {
                this.handleRecurringEventandIncentiveSubmit();
            }
        }
    };

    handleMessageModalClose = () => {
        this.setState(() => ({
            messageOpen: false
        }));
    };

    goToChannel = () => {
        const channel = this.props.post.channel;
        this.props.history.push(`/channels/${channel}`);
    };

    moveToBulletinBoards = () => {
        this.props.history.push("/myChannels");
    };

    returnToPost = () => {
        this.props.history.push(`/post/${this.props.post.post_id}`);
    };

    render() {
        const read_only = !!this.props.post && this.props.post.deleted_flag;

        return (
            <div>
                <Box bgcolor="secondary.main" py={3}>
                    <Container maxWidth="xl">
                        {this.state.step === "Event" && (
                            <Typography variant="h1" gutterBottom>
                                Add an Event to Post:{" "}
                                <Typography variant="inherit" display="inline" color="primary">
                                    {this.props.post && this.props.post.post_title}
                                </Typography>
                            </Typography>
                        )}
                        <CustomStepper steps={this.state.steps} activeStep={this.state.activeStep} />
                        {this.state.step === "Incentive" && (
                            <Typography variant="h1" gutterBottom color="primary">
                                (Optional){" "}
                                <Typography variant="inherit" display="inline" color="textPrimary">
                                    Add Perks to the Event{" "}
                                </Typography>
                            </Typography>
                        )}
                        {read_only ? (
                            <Box>
                                <Typography variant="h2" color="error" gutterBottom>
                                    You must restore the post of this event before adding.
                                </Typography>
                                <Button color="primary" variant="contained" onClick={this.goToPost}>
                                    Go to Post
                                </Button>
                            </Box>
                        ) : (
                            <Box>
                                <span>
                                    {this.state.step === "Incentive" && (
                                        <React.Fragment>
                                            <Button color="primary" variant="contained" onClick={this.toggleForms}>
                                                Go Back
                                            </Button>{" "}
                                        </React.Fragment>
                                    )}
                                    <Button color="primary" variant="contained" onClick={this.goBack}>
                                        See All Post Events
                                    </Button>{" "}
                                    <Button color="primary" variant="contained" onClick={this.recurringOpen}>
                                        Open Recurring Settings
                                    </Button>{" "}
                                </span>
                            </Box>
                        )}
                    </Container>
                </Box>
                <Container maxWidth="xl">
                    <RecurringModal isOpen={this.state.recurringOpen} onSubmit={this.recurringSubmit} />
                    <MessageModal
                        isOpen={this.state.messageOpen}
                        message={this.state.message}
                        closeMessageModal={this.handleMessageModalClose}
                    />
                    {!read_only && (
                        <Box py={2}>
                            {!!this.state.base_incentive && (
                                <React.Fragment>
                                    <Button color="primary" variant="contained" onClick={this.clearBaseIncentive}>
                                        Clear Perks
                                    </Button>{" "}
                                </React.Fragment>
                            )}
                            {!!this.state.base_event && this.state.step === "Event" && (
                                <React.Fragment>
                                    <Button color="primary" variant="contained" onClick={this.clearBaseEvent}>
                                        Clear Event
                                    </Button>{" "}
                                </React.Fragment>
                            )}
                            {!!this.state.base_event && !!this.state.base_incentive && (
                                <React.Fragment>
                                    <Button
                                        color="primary"
                                        variant="contained"
                                        onClick={this.clearBaseEventandIncentive}
                                    >
                                        Clear Event and Perks
                                    </Button>{" "}
                                </React.Fragment>
                            )}
                        </Box>
                    )}
                    <Box>
                        <Typography variant="h3" color="primary">
                            {this.mapRRulesToText()}
                        </Typography>
                    </Box>
                    {!!this.state.event_error && (
                        <Box py={2}>
                            <Typography variant="h3" color="error">
                                {this.state.event_error}
                            </Typography>
                        </Box>
                    )}
                    {!!this.state.incentive_error && (
                        <Box py={2}>
                            <Typography variant="h3" color="error">
                                {this.state.incentive_error}
                            </Typography>
                        </Box>
                    )}
                    {!!this.state.error && (
                        <Box py={2}>
                            <Typography variant="h3" color="error">
                                {this.state.error}
                            </Typography>
                        </Box>
                    )}
                    {this.state.step === "Event" && (
                        <EventForm
                            event={this.state.base_event}
                            read_only={read_only}
                            id="Event"
                            post={this.props.match.params.id}
                            description={!!this.props.post ? this.props.post.post_description : ""}
                            onSubmit={this.onSubmit}
                            nextStep="Continue"
                            showAttendees={false}
                        />
                    )}
                    {this.state.step === "Incentive" && (
                        <IncentiveForm
                            id="Incentive"
                            description={!!this.props.post ? this.props.post.post_description : ""}
                            incentivePackage={this.state.base_incentive}
                            onSubmit={this.onSubmit}
                            nextStep="Save Perks and Go Back"
                            event={this.state.base_event}
                            fromEvent={true}
                        />
                    )}

                    {!read_only && !!this.state.base_event && (
                        <React.Fragment>
                            <Button color="primary" variant="contained" onClick={this.submitAll}>
                                Save and Submit All
                            </Button>{" "}
                        </React.Fragment>
                    )}

                    {this.state.step === "Incentive" && (
                        <React.Fragment>
                            <Button color="primary" variant="contained" onClick={this.submitWithoutSaving}>
                                Skip Perks and Submit
                            </Button>
                        </React.Fragment>
                    )}

                    <button
                        className="button__invisible"
                        type="submit"
                        form={this.state.step}
                        ref={(node) => {
                            this.submitButtonRef = node;
                        }}
                    />
                </Container>
            </div>
        );
    }
}

const mapStateToprops = (state) => ({
    post: state.posts.posts[0]
});

const mapDispatchToProps = (dispatch) => ({
    clearPosts: () => dispatch(clearPosts()),
    addEvent: (event) => dispatch(addEvent(event)),
    addIncentivePackage: (incentive) => dispatch(addIncentivePackage(incentive)),
    startGetPost: (id) => dispatch(startGetPost(id))
});

export default connect(
    mapStateToprops,
    mapDispatchToProps
)(AddEventPage);
