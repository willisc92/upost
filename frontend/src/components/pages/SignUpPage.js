import React from "react";
import { connect } from "react-redux";
import AccountForm from "../forms/AccountForm.js";
import moment from "moment";
import emailValidator from "email-validator";
import { authFail, authSignup } from "../../actions/auth";

class SignUpPage extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            calendarFocused: false,
            date: moment(),
            birthDate: moment(),
            isLoaded: false,
            error: undefined,
            minPasswordLength: 8
        };
    }

    isEmpty = (obj) => {
        for (let key in obj) {
            if (obj.hasOwnProperty(key)) {
                return false;
            }
        }
        return true;
    };

    onSubmit = (input) => {
        let error = {};

        const formInput = input.formInput;

        // validate email
        if (!emailValidator.validate(formInput.email.value)) {
            error.email = ["Invalid email entered"];
        }

        // check password
        if (formInput.password.value !== formInput.passwordConfirmation.value) {
            error.password = ["Passwords do not match"];
        } else if (formInput.password.value.length < this.state.minPasswordLength) {
            error.password = ["Password does not meet minimum length"];
        }

        // check for blanks
        if (formInput.username.value.length === 0) {
            error.username = ["Field cannot be blank."];
        }
        if (formInput.firstName.value.length === 0) {
            error.first_name = ["Field cannot be blank."];
        }
        if (formInput.lastName.value.length === 0) {
            error.last_name = ["Field cannot be blank."];
        }
        if (formInput.country.value.length === 0) {
            error.country = ["Field cannot be blank."];
        }
        if (formInput.state.value.length === 0) {
            error.state = ["Field cannot be blank."];
        }
        if (formInput.streetName.value.length === 0) {
            error.street_name = ["Field cannot be blank."];
        }
        if (formInput.postalCode.value.length === 0) {
            error.postal_code = ["Field cannot be blank."];
        }
        if (formInput.city.value.length === 0) {
            error.city = ["Field cannot be blank."];
        }

        if (!this.isEmpty(error)) {
            console.log("client side error", error);
            this.props.authFail(error);
            return;
        }

        const user = {
            password: formInput.password.value,
            username: formInput.username.value,
            email: formInput.email.value,
            first_name: formInput.firstName.value,
            middle_name: formInput.middleName.value,
            last_name: formInput.lastName.value,
            birth_date: input.date.format("YYYY-MM-DD"),
            country: formInput.country.value,
            state: formInput.state.value,
            street_name: formInput.streetName.value,
            postal_code: formInput.postalCode.value,
            city: formInput.city.value,
            sex: formInput.sex.value === "Male" ? "M" : "F",
            phone_number: formInput.phoneNumber.value,
            channels: [] // not ideal, ideally don't include fields we dont need have django handle
        };

        this.props
            .authSignup(user)
            .then(() => {
                this.props.history.push("/interests");
            })
            .catch((error) => {
                console.log("An error has occured with signup", error);
            });
    };

    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page_header__title"> SignUp </h1>
                    </div>
                </div>
                <div className="content-container">
                    <AccountForm onSubmit={this.onSubmit} minPasswordLength={this.state.minPasswordLength} />
                </div>
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        authFail: (error) => dispatch(authFail(error)),
        authSignup: (user) => dispatch(authSignup(user))
    };
};

export default connect(
    undefined,
    mapDispatchToProps
)(SignUpPage);
