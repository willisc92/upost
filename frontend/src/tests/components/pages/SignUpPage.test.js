import { SignUpPage } from "../../../components/pages/SignUpPage";
import React from "react";
import { shallow } from "enzyme";
import moment from "moment";

let wrapper, authSignup, authFail, history;

beforeEach(() => {
    authSignup = jest.fn().mockImplementation(() => {
        return Promise.resolve();
    });
    history = {
        push: jest.fn()
    };
    authFail = jest.fn();
    wrapper = shallow(<SignUpPage authSignup={authSignup} authFail={authFail} history={history} />);
});

test("Should render SignUpPage", () => {
    expect(wrapper).toMatchSnapshot();
});

test("Should handle onSubmit", async () => {
    const input = {
        date: moment(),
        formInput: {
            email: {
                value: "testemail@hotmail.com"
            },
            username: {
                value: "testuser"
            },
            password: {
                value: "test_password"
            },
            passwordConfirmation: {
                value: "test_password"
            },
            firstName: {
                value: "first_name"
            },
            middleName: {
                value: "middle_name"
            },
            lastName: {
                value: "last_name"
            },
            country: {
                value: "country"
            },
            state: {
                value: "state"
            },
            streetName: {
                value: "123 fake street"
            },
            postalCode: {
                value: "A1B 2R1"
            },
            city: {
                value: "Calgary"
            },
            sex: {
                value: "Male"
            },
            phoneNumber: {
                value: 4034010115
            }
        }
    };

    await wrapper.instance().onSubmit(input);
    expect(history.push).toHaveBeenCalledWith("/interests");
});
