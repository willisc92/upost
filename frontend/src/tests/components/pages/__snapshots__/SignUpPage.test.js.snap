// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render SignUpPage 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page_header__title"
      >
         SignUp 
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <Connect(AccountForm)
      minPasswordLength={8}
      onSubmit={[Function]}
    />
  </div>
</div>
`;
